server/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -389,6 +390,40 @@ public void onStoreClosed(ShardId shardId) {
         }
     }
 
+    public <T, E extends Exception> T withTempIndexService(final IndexMetadata indexMetadata,
+                                                           CheckedFunction<IndexService, T, E> indexServiceConsumer) throws IOException, E {
+        final Index index = indexMetadata.getIndex();
+        if (hasIndex(index)) {
+            throw new ResourceAlreadyExistsException(index);
+        }
+        List<IndexEventListener> finalListeners = Collections.singletonList(
+            // double check that shard is not created.
+            new IndexEventListener() {
+                @Override
+                public void beforeIndexShardCreated(ShardId shardId, Settings indexSettings) {
+                    assert false : "temp index should not trigger shard creation";
+                    throw new ElasticsearchException("temp index should not trigger shard creation [{}]", index);
+                }
+
+                @Override
+                public void onStoreCreated(ShardId shardId) {
+                    assert false : "temp index should not trigger store creation";
+                    throw new ElasticsearchException("temp index should not trigger store creation [{}]", index);
+                }
+            }
+        );
+        final IndexService indexService =
+            createIndexService(
+                IndexCreationContext.CREATE_INDEX,
+                indexMetadata,
+                indicesQueryCache,
+                finalListeners,
+                indexingMemoryController);
+        try (Closeable dummy = () -> indexService.close("temp", false)) {
```

Comment:
- So this now closes the index if something fails and it's in-between state, right?
- Is it ok that we loose this extra info we had instead of `"temp"`? From a quick look I don't see that something was logged anyway, just double checking, in case I miss something.

Reply:
If you think it can help debugging some issue, I'd say let's keep it, It's not that complex after all to pass them right?

Reply:
I'm not sure if they're really helpful - that's why I left them out.